File: backend/services/google_sheets.py
import requests
import logging
from datetime import datetime
from config import Config

logger = logging.getLogger(__name__)

class GoogleSheetsService:
    def __init__(self):
        self.api_key = Config.GOOGLE_SHEETS_API_KEY
        self.spreadsheet_id = Config.GOOGLE_SHEETS_ID
        self.sheet_name = Config.GOOGLE_SHEETS_TAB
        self.users_sheet_name = 'Users'
        self.base_url = 'https://sheets.googleapis.com/v4/spreadsheets'

    def initialize_sheet(self):
        """Initialize the spreadsheet with headers"""
        headers = [
            'Timestamp',
            'Event ID',
            'Severity',
            'Source IP',
            'Destination IP',
            'Attack Type',
            'Status',
            'Action Taken'
        ]

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}!A1:H1"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': [headers]}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {'success': True, 'data': response.json()}
        else:
            logger.error(f"Failed to initialize sheet: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to initialize sheet')}

    def add_security_incident(self, incident):
        """Add a new security incident"""
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        event_id = incident.get('eventId') or f"INC-{datetime.now().strftime('%Y%m%d')}-{str(hash(str(incident)) % 1000).zfill(3)}"

        row_data = [
            timestamp,
            event_id,
            incident.get('severity', 'Medium'),
            incident.get('sourceip', 'Unknown'),
            incident.get('destinationip', 'Unknown'),
            incident.get('attacktype', 'Unknown'),
            incident.get('status', 'Investigating'),
            incident.get('actiontaken', 'Alert Sent')
        ]

        # Get current row count
        range_url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}"
        response = requests.get(range_url, params={'key': self.api_key})
        if response.status_code != 200:
            return {'success': False, 'error': 'Failed to get sheet data'}

        current_rows = len(response.json().get('values', []))
        next_row = current_rows + 1

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}!A{next_row}:H{next_row}"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': [row_data]}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {
                'success': True,
                'data': {
                    'incident': {
                        'timestamp': timestamp,
                        'eventId': event_id,
                        'severity': incident.get('severity'),
                        'sourceIp': incident.get('sourceIp'),
                        'destinationIp': incident.get('destinationIp'),
                        'attackType': incident.get('attackType'),
                        'status': incident.get('status'),
                        'actionTaken': incident.get('actionTaken')
                    }
                }
            }
        else:
            logger.error(f"Failed to add incident: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to add incident')}

    def get_security_incidents(self):
        """Get all security incidents"""
        csv_url = f"https://docs.google.com/spreadsheets/d/{self.spreadsheet_id}/export?format=csv"
        response = requests.get(csv_url)
        if response.status_code != 200:
            return {'success': False, 'error': 'Failed to fetch CSV data'}

        lines = response.text.strip().split('\n')
        if len(lines) <= 1:
            return {'success': True, 'data': []}

        header_row = lines[0].split(',')
        incidents = []
        for line in lines[1:]:
            values = line.split(',')
            incident = {}
            for i, header in enumerate(header_row):
                incident[header.lower().replace(' ', '')] = values[i] if i < len(values) else ''
            incidents.append(incident)

        return {'success': True, 'data': incidents}

    def update_incident_status(self, event_id, new_status, action_taken):
        """Update incident status"""
        incidents = self.get_security_incidents()
        if not incidents['success']:
            return incidents

        incident_index = None
        for i, inc in enumerate(incidents['data']):
            if inc.get('eventid') == event_id:
                incident_index = i
                break

        if incident_index is None:
            return {'success': False, 'error': 'Incident not found'}

        row_number = incident_index + 2  # +1 for header, +1 for 1-based

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}!G{row_number}:H{row_number}"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': [[new_status, action_taken]]}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {'success': True, 'data': response.json()}
        else:
            logger.error(f"Failed to update incident: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to update incident')}

    def bulk_add_incidents(self, incidents_list):
        """Bulk add multiple incidents"""
        rows = []
        for incident in incidents_list:
            timestamp = incident.get('timestamp') or datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            event_id = incident.get('eventId') or f"INC-{datetime.now().strftime('%Y%m%d')}-{str(hash(str(incident)) % 1000).zfill(3)}"

            rows.append([
                timestamp,
                event_id,
                incident.get('severity', 'Medium'),
                incident.get('sourceIp', 'Unknown'),
                incident.get('destinationIp', 'Unknown'),
                incident.get('attackType', 'Unknown'),
                incident.get('status', 'Investigating'),
                incident.get('actionTaken', 'Alert Sent')
            ])

        # Get current row count
        range_url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}"
        response = requests.get(range_url, params={'key': self.api_key})
        if response.status_code != 200:
            return {'success': False, 'error': 'Failed to get sheet data'}

        current_rows = len(response.json().get('values', []))
        start_row = current_rows + 1
        end_row = start_row + len(rows) - 1

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.sheet_name}!A{start_row}:H{end_row}"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': rows}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {'success': True, 'data': {'addedCount': len(rows)}}
        else:
            logger.error(f"Failed to bulk add incidents: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to bulk add incidents')}

    def initialize_users_sheet(self):
        """Initialize the users sheet with headers"""
        headers = ['Email', 'Role', 'Created At']

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.users_sheet_name}!A1:C1"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': [headers]}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {'success': True, 'data': response.json()}
        else:
            logger.error(f"Failed to initialize users sheet: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to initialize users sheet')}

    def add_user(self, email, role):
        """Add a new user"""
        created_at = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

        row_data = [email, role, created_at]

        # Get current row count
        range_url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.users_sheet_name}"
        response = requests.get(range_url, params={'key': self.api_key})
        if response.status_code != 200:
            # Try to initialize the sheet
            init_result = self.initialize_users_sheet()
            if not init_result['success']:
                return {'success': False, 'error': 'Failed to initialize users sheet'}
            # Retry get
            response = requests.get(range_url, params={'key': self.api_key})
            if response.status_code != 200:
                return {'success': False, 'error': 'Failed to get users sheet data after initialization'}

        current_rows = len(response.json().get('values', []))
        next_row = current_rows + 1

        url = f"{self.base_url}/{self.spreadsheet_id}/values/{self.users_sheet_name}!A{next_row}:C{next_row}"
        params = {'key': self.api_key, 'valueInputOption': 'RAW'}
        data = {'values': [row_data]}

        response = requests.put(url, json=data, params=params)
        if response.status_code == 200:
            return {'success': True, 'data': {'email': email, 'role': role, 'createdAt': created_at}}
        else:
            logger.error(f"Failed to add user: {response.text}")
            return {'success': False, 'error': response.json().get('error', {}).get('message', 'Failed to add user')}

    def get_user_by_email(self, email):
        """Get user by email"""
        csv_url = f"https://docs.google.com/spreadsheets/d/{self.spreadsheet_id}/export?format=csv&sheet={self.users_sheet_name}"
        response = requests.get(csv_url)
        if response.status_code != 200:
            return {'success': False, 'error': 'Failed to fetch users CSV data'}

        lines = response.text.strip().split('\n')
        if len(lines) <= 1:
            return {'success': True, 'data': None}

        header_row = lines[0].split(',')
        for line in lines[1:]:
            values = line.split(',')
            if len(values) >= 2 and values[0] == email:
                user = {}
                for i, header in enumerate(header_row):
                    user[header.lower().replace(' ', '')] = values[i] if i < len(values) else ''
                return {'success': True, 'data': user}

        return {'success': True, 'data': None}

File: src/services/googleSheets.js
import axios from 'axios';
import { getGoogleToken } from './googleAuth';

class GoogleSheetsService {
  constructor() {
    this.apiKey = import.meta.env?.VITE_GOOGLE_SHEETS_API_KEY;
    this.spreadsheetId = import.meta.env?.VITE_GOOGLE_SHEETS_ID || '1P988HAfB1GGLmNqXv6cnA35EBbgvJSViQrE3EnU-Y9o';
    this.sheetName = import.meta.env?.VITE_GOOGLE_SHEETS_TAB || 'Sheet1';
    this.baseUrl = 'https://sheets.googleapis.com/v4/spreadsheets';
    
    console.log('GoogleSheetsService initialized with:');
    console.log('- API Key:', this.apiKey ? 'Set' : 'Not set');
    console.log('- Spreadsheet ID:', this.spreadsheetId);
    console.log('- Sheet Name:', this.sheetName);
  }

  /**
   * Initialize the spreadsheet with headers
   */
  async initializeSheet() {
    console.log('Initializing Google Sheet with headers...');
    try {
      const headers = [
        'Timestamp',
        'Event ID', 
        'Severity',
        'Source IP',
        'Destination IP',
        'Attack Type',
        'Status',
        'Action Taken'
      ];
      
      console.log('Headers to initialize:', headers);

      const response = await axios?.put(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}!A1:H1`,
        {
          values: [headers]
        },
        {
          params: {
            key: this.apiKey,
            valueInputOption: 'RAW'
          },
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );

      console.log('Sheet initialization successful:', response?.data);
      return {
        success: true,
        data: response?.data
      };
    } catch (error) {
      console.error('Google Sheets initialization error:', error);
      console.error('Error details:', error?.response?.data?.error || error.message);
      return {
        success: false,
        error: error?.response?.data?.error?.message || 'Failed to initialize sheet'
      };
    }
  }

  /**
   * Add a new security incident to the Google Sheet
   * @param {Object} incident - Security incident data
   * @returns {Object} Response from Google Sheets API
   */
  async addSecurityIncident(incident) {
    console.log('Adding security incident to Google Sheet:', incident);
    try {
      const timestamp = new Date()?.toLocaleString('en-US', {
        year: 'numeric',
        month: '2-digit', 
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        hour12: false
      });

      const eventId = incident?.eventId || `INC-${new Date()?.toISOString()?.split('T')?.[0]?.replace(/-/g, '')}-${String(Math.floor(Math.random() * 1000))?.padStart(3, '0')}`;

      const rowData = [
        timestamp,
        eventId,
        incident?.severity || 'Medium',
        incident?.sourceIp || 'Unknown',
        incident?.destinationIp || 'Unknown', 
        incident?.attackType || 'Unknown',
        incident?.status || 'Investigating',
        incident?.actionTaken || 'Alert Sent'
      ];
      
      console.log('Row data to add:', rowData);

      // Get current row count to append to next available row
      console.log('Getting current row count...');
      const rangeResponse = await axios?.get(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}`,
        {
          params: {
            key: this.apiKey
          }
        }
      );
      
      console.log('Current sheet data:', rangeResponse?.data);

      const currentRows = rangeResponse?.data?.values?.length || 1;
      const nextRow = currentRows + 1;
      console.log(`Adding incident to row ${nextRow}`);

      const response = await axios?.put(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}!A${nextRow}:H${nextRow}`,
        {
          values: [rowData]
        },
        {
          params: {
            key: this.apiKey,
            valueInputOption: 'RAW'
          },
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );

      console.log('Incident added successfully:', response?.data);
      return {
        success: true,
        data: {
          ...response?.data,
          incident: {
            timestamp,
            eventId,
            severity: incident?.severity,
            sourceIp: incident?.sourceIp,
            destinationIp: incident?.destinationIp,
            attackType: incident?.attackType,
            status: incident?.status,
            actionTaken: incident?.actionTaken
          }
        }
      };
    } catch (error) {
      console.error('Google Sheets add incident error:', error);
      console.error('Error details:', error?.response?.data?.error || error.message);
      return {
        success: false,
        error: error?.response?.data?.error?.message || 'Failed to add incident to sheet'
      };
    }
  }

  /**
   * Get all security incidents from the Google Sheet
   * @returns {Array} Array of security incidents
   */
  async getSecurityIncidents() {
    console.log('Fetching security incidents from public CSV URL...');
    try {
      const csvUrl = 'https://docs.google.com/spreadsheets/d/1P988HAfB1GGLmNqXv6cnA35EBbgvJSViQrE3EnU-Y9o/export?format=csv';
      const response = await fetch(csvUrl);
      if (!response.ok) {
        throw new Error(`Failed to fetch CSV data: ${response.status} ${response.statusText}`);
      }
      const csvText = await response.text();

      // Parse CSV text
      const lines = csvText.trim().split('\n');
      if (lines.length <= 1) {
        console.log('No incident data found in CSV (only headers or empty)');
        return {
          success: true,
          data: []
        };
      }

      const headerRow = lines[0].split(',');
      console.log('Headers from CSV:', headerRow);

      const incidents = lines.slice(1).map(line => {
        const values = line.split(',');
        const incident = {};
        headerRow.forEach((header, index) => {
          incident[header.toLowerCase().replace(/\s+/g, '')] = values[index] || '';
        });
        return incident;
      });

      console.log(`Processed ${incidents.length} incidents from CSV:`, incidents);

      return {
        success: true,
        data: incidents
      };
    } catch (error) {
      console.error('CSV fetch get incidents error:', error);
      return {
        success: false,
        error: error.message || 'Failed to fetch incidents from CSV'
      };
    }
  }

  /**
   * Update incident status in Google Sheet
   * @param {string} eventId - Event ID to update
   * @param {string} newStatus - New status
   * @param {string} actionTaken - Action taken
   * @returns {Object} Update response
   */
  async updateIncidentStatus(eventId, newStatus, actionTaken) {
    console.log(`Updating incident ${eventId} with status: ${newStatus}, action: ${actionTaken}`);
    try {
      // First, get all incidents to find the row
      console.log('Fetching all incidents to locate the target row...');
      const incidents = await this.getSecurityIncidents();
      
      if (!incidents?.success) {
        console.error('Failed to fetch incidents for update:', incidents.error);
        return incidents;
      }

      const incidentIndex = incidents?.data?.findIndex(
        incident => incident?.eventid === eventId
      );
      
      console.log(`Incident index for ${eventId}: ${incidentIndex}`);

      if (incidentIndex === -1) {
        console.error(`Incident with ID ${eventId} not found`);
        return {
          success: false,
          error: 'Incident not found'
        };
      }

      // Update the specific cells (row index + 2 to account for header and 0-based indexing)
      const rowNumber = incidentIndex + 2;
      console.log(`Updating row ${rowNumber} with new status and action`);
      
      const updateResponse = await axios?.put(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}!G${rowNumber}:H${rowNumber}`,
        {
          values: [[newStatus, actionTaken]]
        },
        {
          params: {
            key: this.apiKey,
            valueInputOption: 'RAW'
          },
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
      
      console.log('Update response:', updateResponse?.data);

      return {
        success: true,
        data: updateResponse?.data
      };
    } catch (error) {
      console.error('Google Sheets update error:', error);
      console.error('Error details:', error?.response?.data?.error || error.message);
      return {
        success: false,
        error: error?.response?.data?.error?.message || 'Failed to update incident'
      };
    }
  }

  /**
   * Bulk add multiple incidents
   * @param {Array} incidents - Array of incident objects
   * @returns {Object} Bulk add response
   */
  async bulkAddIncidents(incidents) {
    console.log(`Bulk adding ${incidents?.length} incidents to Google Sheet`);
    try {
      const rows = incidents?.map(incident => {
        const timestamp = incident?.timestamp || new Date()?.toLocaleString('en-US', {
          year: 'numeric',
          month: '2-digit', 
          day: '2-digit',
          hour: '2-digit',
          minute: '2-digit',
          hour12: false
        });

        const eventId = incident?.eventId || `INC-${new Date()?.toISOString()?.split('T')?.[0]?.replace(/-/g, '')}-${String(Math.floor(Math.random() * 1000))?.padStart(3, '0')}`;

        return [
          timestamp,
          eventId,
          incident?.severity || 'Medium',
          incident?.sourceIp || 'Unknown',
          incident?.destinationIp || 'Unknown',
          incident?.attackType || 'Unknown',
          incident?.status || 'Investigating',
          incident?.actionTaken || 'Alert Sent'
        ];
      });
      
      console.log('Prepared rows for bulk add:', rows);

      // Get current row count
      console.log('Getting current row count...');
      const rangeResponse = await axios?.get(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}`,
        {
          params: {
            key: this.apiKey
          }
        }
      );

      const currentRows = rangeResponse?.data?.values?.length || 1;
      const startRow = currentRows + 1;
      const endRow = startRow + rows?.length - 1;
      
      console.log(`Adding incidents to rows ${startRow}-${endRow}`);

      const response = await axios?.put(
        `${this.baseUrl}/${this.spreadsheetId}/values/${this.sheetName}!A${startRow}:H${endRow}`,
        {
          values: rows
        },
        {
          params: {
            key: this.apiKey,
            valueInputOption: 'RAW'
          },
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
      
      console.log('Bulk add response:', response?.data);

      return {
        success: true,
        data: {
          ...response?.data,
          addedCount: rows?.length
        }
      };
    } catch (error) {
      console.error('Google Sheets bulk add error:', error);
      console.error('Error details:', error?.response?.data?.error || error.message);
      return {
        success: false,
        error: error?.response?.data?.error?.message || 'Failed to bulk add incidents'
      };
    }
  }
}

export default new GoogleSheetsService();